### DISCUSSION TITLE
Many to many internal implementation

### DISCUSSION THREAD

**Discussion by tha23rd** [pocketbase/pocketbase#2809] [Q&A] [answered]
Hi! First off, just want to say that this is an awesome open source project and I'm super appreciative of all of the work you and others have put in

For a recent side project I needed to implement a many-to-many relationship. I noticed that internally, this was being represented using String types in both tables which contain a "list" of the ids that can be mapped to.

I was wondering why this approach was taken vs using a join table. For certain queries I would like to make (which admittedly are more analytics focused), this implementation results in extremely long query times. I'm sure there were trade offs made that I'm not aware of that led to this, but am just curious.

Thanks!

**Comment by ganigeorgiev** [comment, marked as answer]
> I was wondering why this approach was taken vs using a join table.

You have both approaches available. Nothing is stopping you to create a junction collection with 2 single `relation` fields, the same way as in a more traditional SQL structure.

**Reply by tha23rd** [reply]
Got it, will explore refactoring using that approach - appreciate the response!

**Reply by ganigeorgiev** [reply]
Also keep in mind that the `relation` fields don't create indexes by default (_there is no foreign key constriaint as the relation could be "dynamic"_) and in case your query doesn't perform well you may want to create indexes for the related fields (index expressions are also supported).

**Reply by tha23rd** [reply]
I see - Sorry this is going a bit beyond my SQL experience. To read that back to you:

In the join table, where each row defines a relationship between table "a" and table "b", I should create indices on each column? For context, one query pattern I'm trying to support is "sum the number of times "b" is present in an "a" over the last x days for each b"

**Reply by ganigeorgiev** [reply]
> In the join table, where each row defines a relationship between table "a" and table "b", I should create indices on each column? 

It is not required, but Yes, if the performance suffers because of it you can try to add 2 separate indexes from the Admin UI (one for each `relation` field) and see if that will fix the slow query times.

Note that even with indexes on the `relation` fields, it is not guaranteed that the query planner will use them as this will differ from one query to another. It requires a little "trial and error" approach to fine tune the query. You can enable the query std logs with  `--debug`.

To see what route the SQLite query planner takes, you can copy the slow query from the console log (_replace some of its placeholder values_) and run it with `EXPLAIN YOUR_QUERY` or (`EXPLAIN QUERY PLAN YOUR_QUERY`) in the [`sqlite3 cli` tool](https://www.sqlite.org/cli.html) (_probably installed by default on most platforms_) and adjust accordingly.

**Reply by tha23rd** [reply]
Awesome, thanks a ton for all of the help/advice

**Reply by ganigeorgiev** [reply]
Additionally, if even with the appropiate indexes your query doesn't perform as good as you want it, you can try to build PocketBase with `CGO_ENABLED=1` in order to kick in the cgo SQLite driver (for some queries it could be ~2x+ faster than the pure Go SQLite port that we use by default; there will be official testbed and benchmarks soon). For build instructions, you can check https://github.com/pocketbase/pocketbase#running-and-building.

**Reply by tha23rd** [reply]
Just wanted to report back here -

After migrating my data to use the new schema utilizing junction tables, performance is now significantly better. Creating indices as mentioned above on the junction also improved performance by another ~4x

Thanks again!

**Reply by ganigeorgiev** [reply]
While working on the benchmarks I also noticed that replacing the `COUNT(DISTINCT table.id)` with `COUNT(DISTINCT table._rowid_)` improves the count performance for large datasets (150k) almost 10x (from ~580ms to ~60ms).

I'll consider pushing a v0.16.7 release sometime later today.

**Reply by tha23rd** [reply]
I wonder if that also extends to `GROUP BY`? I.e `GROUP BY table.id` vs `GROUP BY table._rowid_`?

**Reply by ganigeorgiev** [reply]
I'm not sure, it is difficult to say as it will depend on the final query. The best approach would be to test it and/or analyze the query plan with `EXPLAIN QUERY PLAN YOUR_QUERY` in the sqlite3 cli tool. There is also [`.expert` command](https://www.sqlite.org/cli.html#index_recommendations_sqlite_expert_) that can provide some suggestions/recommendations what indexes to add to the query (_I have it in my todo to research whether something like this eventually can be added into PocketBase directly_).
